File: server/hud/runtime.py
from __future__ import annotations

import asyncio
from datetime import datetime, timezone

from .adapters.base import CommandCodeAdapter, PublicAdapterError, ZCodeAdapter
from .config import RuntimeConfig
from .models import CommandCodeState, SourceHealth, ZCodeState
from .store import SnapshotStore


def _utcnow() -> datetime:
    return datetime.now(timezone.utc)


def _public_error_message(error: Exception) -> str:
    if isinstance(error, PublicAdapterError):
        message = str(error).strip() or "source collection failed"
        return message[:240]
    return f"{type(error).__name__}: source collection failed"[:240]


class HudRuntime:
    """Run source collectors independently and commit normalized state to the store."""

    def __init__(
        self,
        store: SnapshotStore,
        *,
        zcode_adapter: ZCodeAdapter | None = None,
        command_code_adapter: CommandCodeAdapter | None = None,
        config: RuntimeConfig | None = None,
    ):
        self.store = store
        self.zcode_adapter = zcode_adapter
        self.command_code_adapter = command_code_adapter
        self.config = config or RuntimeConfig()
        self._tasks: list[asyncio.Task[None]] = []

    async def start(self) -> None:
        if self._tasks:
            return
        if self.zcode_adapter is not None:
            self._tasks.append(
                asyncio.create_task(self._zcode_loop(), name="hud-zcode-collector")
            )
        if self.command_code_adapter is not None:
            self._tasks.append(
                asyncio.create_task(
                    self._command_code_loop(), name="hud-commandcode-collector"
                )
            )

    async def stop(self) -> None:
        tasks, self._tasks = self._tasks, []
        for task in tasks:
            task.cancel()
        if tasks:
            await asyncio.gather(*tasks, return_exceptions=True)

    async def collect_zcode_once(self) -> None:
        if self.zcode_adapter is None:
            return
        try:
            payload = await self.zcode_adapter.collect()
            observed_at = _utcnow()
            state = ZCodeState(
                health=SourceHealth(
                    status="ok",
                    observed_at=observed_at,
                    last_success_at=observed_at,
                    message=None,
                ),
                summary=payload.summary,
                tasks=payload.tasks,
            )
        except Exception as error:
            await self._record_zcode_failure(_utcnow(), error)
            return
        await self.store.update_zcode(state)

    async def collect_command_code_once(self) -> None:
        if self.command_code_adapter is None:
            return
        try:
            payload = await self.command_code_adapter.collect()
            observed_at = _utcnow()
            state = CommandCodeState(
                health=SourceHealth(
                    status="ok",
                    observed_at=observed_at,
                    last_success_at=observed_at,
                    message=None,
                ),
                usage=payload.usage,
            )
        except Exception as error:
            await self._record_command_code_failure(_utcnow(), error)
            return
        await self.store.update_command_code(state)

    async def _record_zcode_failure(
        self, observed_at: datetime, error: Exception
    ) -> None:
        previous = (await self.store.get()).zcode
        has_last_good = (
            previous.summary is not None
            and previous.tasks is not None
            and previous.health.last_success_at is not None
        )
        if has_last_good:
            state = ZCodeState(
                health=SourceHealth(
                    status="stale",
                    observed_at=observed_at,
                    last_success_at=previous.health.last_success_at,
                    message=_public_error_message(error),
                ),
                summary=previous.summary,
                tasks=previous.tasks,
            )
        else:
            state = ZCodeState(
                health=SourceHealth(
                    status="error",
                    observed_at=observed_at,
                    last_success_at=None,
                    message=_public_error_message(error),
                ),
                summary=None,
                tasks=None,
            )
        await self.store.update_zcode(state)

    async def _record_command_code_failure(
        self, observed_at: datetime, error: Exception
    ) -> None:
        previous = (await self.store.get()).command_code
        has_last_good = (
            previous.usage is not None and previous.health.last_success_at is not None
        )
        if has_last_good:
            state = CommandCodeState(
                health=SourceHealth(
                    status="stale",
                    observed_at=observed_at,
                    last_success_at=previous.health.last_success_at,
                    message=_public_error_message(error),
                ),
                usage=previous.usage,
            )
        else:
            state = CommandCodeState(
                health=SourceHealth(
                    status="error",
                    observed_at=observed_at,
                    last_success_at=None,
                    message=_public_error_message(error),
                ),
                usage=None,
            )
        await self.store.update_command_code(state)

    async def _zcode_loop(self) -> None:
        while True:
            await self.collect_zcode_once()
            await asyncio.sleep(self.config.zcode_interval_seconds)

    async def _command_code_loop(self) -> None:
        while True:
            await self.collect_command_code_once()
            await asyncio.sleep(self.config.command_code_interval_seconds)

File: agent/internal/runtime/runtime.go
package runtime

import (
	"context"
	"fmt"
	"strings"
	"sync"
	"time"

	"github.com/Eswink/ai-control-hud/agent/internal/collector/zcode"
	"github.com/Eswink/ai-control-hud/agent/internal/domain"
	"github.com/Eswink/ai-control-hud/agent/internal/store"
)

type ZCodeCollectFunc func(context.Context) (*zcode.Snapshot, error)
type CommandCodeCollectFunc func(context.Context) (*domain.UsageSummary, error)

type Config struct {
	ZCodeInterval       time.Duration
	CommandCodeInterval time.Duration
	ZCodeTimeout        time.Duration
	CommandCodeTimeout  time.Duration
}

func DefaultConfig() Config {
	return Config{
		ZCodeInterval:       time.Second,
		CommandCodeInterval: 60 * time.Second,
		ZCodeTimeout:        3 * time.Second,
		CommandCodeTimeout:  10 * time.Second,
	}
}

type Runtime struct {
	store       *store.SnapshotStore
	zcode       ZCodeCollectFunc
	commandCode CommandCodeCollectFunc
	config      Config
	now         func() time.Time

	startOnce sync.Once
	wg        sync.WaitGroup
}

func New(
	store *store.SnapshotStore,
	zcodeCollect ZCodeCollectFunc,
	commandCodeCollect CommandCodeCollectFunc,
	config Config,
) *Runtime {
	defaults := DefaultConfig()
	if config.ZCodeInterval <= 0 {
		config.ZCodeInterval = defaults.ZCodeInterval
	}
	if config.CommandCodeInterval <= 0 {
		config.CommandCodeInterval = defaults.CommandCodeInterval
	}
	if config.ZCodeTimeout <= 0 {
		config.ZCodeTimeout = defaults.ZCodeTimeout
	}
	if config.CommandCodeTimeout <= 0 {
		config.CommandCodeTimeout = defaults.CommandCodeTimeout
	}
	return &Runtime{
		store:       store,
		zcode:       zcodeCollect,
		commandCode: commandCodeCollect,
		config:      config,
		now:         time.Now,
	}
}

func InitialState(now time.Time, version string, zcodeEnabled, commandCodeEnabled bool) domain.HudState {
	now = now.UTC()
	return domain.HudState{
		SchemaVersion: domain.SchemaVersion,
		Server: domain.ServerInfo{
			Version: version,
			Time:    now,
		},
		Overall: domain.OverallStatus{Status: domain.OverallDegraded},
		ZCode: domain.ZCodeState{
			Health: initialHealth(now, zcodeEnabled, "ZCode collector is not configured"),
		},
		CommandCode: domain.CommandCodeState{
			Health: initialHealth(now, commandCodeEnabled, "CommandCode collector is not configured"),
		},
	}
}

func initialHealth(now time.Time, enabled bool, disabledMessage string) domain.SourceHealth {
	if enabled {
		message := "collector is waiting for the first snapshot"
		return domain.SourceHealth{Status: domain.SourceError, ObservedAt: now, Message: &message}
	}
	message := disabledMessage
	return domain.SourceHealth{Status: domain.SourceDisabled, ObservedAt: now, Message: &message}
}

func (r *Runtime) Start(ctx context.Context) {
	r.startOnce.Do(func() {
		if r.zcode != nil {
			r.wg.Add(1)
			go r.zcodeLoop(ctx)
		}
		if r.commandCode != nil {
			r.wg.Add(1)
			go r.commandCodeLoop(ctx)
		}
	})
}

func (r *Runtime) Wait() {
	r.wg.Wait()
}

func (r *Runtime) CollectZCodeOnce(parent context.Context) {
	if r.zcode == nil {
		return
	}
	ctx, cancel := context.WithTimeout(parent, r.config.ZCodeTimeout)
	defer cancel()

	snapshot, err := r.zcode(ctx)
	observedAt := r.utcnow()
	if err != nil || snapshot == nil {
		if err == nil {
			err = fmt.Errorf("ZCode collection returned no snapshot")
		}
		r.recordZCodeFailure(observedAt, err)
		return
	}

	lastSuccess := observedAt
	summary := snapshot.Summary
	tasks := snapshot.Tasks
	if tasks == nil {
		tasks = []domain.TaskSummary{}
	}
	_ = r.store.UpdateZCode(domain.ZCodeState{
		Health: domain.SourceHealth{
			Status:        domain.SourceOK,
			ObservedAt:    observedAt,
			LastSuccessAt: &lastSuccess,
		},
		Summary: &summary,
		Tasks:   tasks,
	})
}

func (r *Runtime) CollectCommandCodeOnce(parent context.Context) {
	if r.commandCode == nil {
		return
	}
	ctx, cancel := context.WithTimeout(parent, r.config.CommandCodeTimeout)
	defer cancel()

	usage, err := r.commandCode(ctx)
	observedAt := r.utcnow()
	if err != nil || usage == nil {
		if err == nil {
			err = fmt.Errorf("CommandCode collection returned no usage")
		}
		r.recordCommandCodeFailure(observedAt, err)
		return
	}

	lastSuccess := observedAt
	_ = r.store.UpdateCommandCode(domain.CommandCodeState{
		Health: domain.SourceHealth{
			Status:        domain.SourceOK,
			ObservedAt:    observedAt,
			LastSuccessAt: &lastSuccess,
		},
		Usage: usage,
	})
}

func (r *Runtime) recordZCodeFailure(observedAt time.Time, err error) {
	previous := r.store.Get().ZCode
	message := publicMessage("zcode", err)
	if previous.Summary != nil && previous.Tasks != nil && previous.Health.LastSuccessAt != nil {
		_ = r.store.UpdateZCode(domain.ZCodeState{
			Health: domain.SourceHealth{
				Status:        domain.SourceStale,
				ObservedAt:    observedAt,
				LastSuccessAt: previous.Health.LastSuccessAt,
				Message:       &message,
			},
			Summary: previous.Summary,
			Tasks:   previous.Tasks,
		})
		return
	}
	_ = r.store.UpdateZCode(domain.ZCodeState{
		Health: domain.SourceHealth{
			Status:     domain.SourceError,
			ObservedAt: observedAt,
			Message:    &message,
		},
	})
}

func (r *Runtime) recordCommandCodeFailure(observedAt time.Time, err error) {
	previous := r.store.Get().CommandCode
	message := publicMessage("commandcode", err)
	if previous.Usage != nil && previous.Health.LastSuccessAt != nil {
		_ = r.store.UpdateCommandCode(domain.CommandCodeState{
			Health: domain.SourceHealth{
				Status:        domain.SourceStale,
				ObservedAt:    observedAt,
				LastSuccessAt: previous.Health.LastSuccessAt,
				Message:       &message,
			},
			Usage: previous.Usage,
		})
		return
	}
	_ = r.store.UpdateCommandCode(domain.CommandCodeState{
		Health: domain.SourceHealth{
			Status:     domain.SourceError,
			ObservedAt: observedAt,
			Message:    &message,
		},
	})
}

func (r *Runtime) zcodeLoop(ctx context.Context) {
	defer r.wg.Done()
	ticker := time.NewTicker(r.config.ZCodeInterval)
	defer ticker.Stop()
	for {
		r.CollectZCodeOnce(ctx)
		select {
		case <-ctx.Done():
			return
		case <-ticker.C:
		}
	}
}

func (r *Runtime) commandCodeLoop(ctx context.Context) {
	defer r.wg.Done()
	ticker := time.NewTicker(r.config.CommandCodeInterval)
	defer ticker.Stop()
	for {
		r.CollectCommandCodeOnce(ctx)
		select {
		case <-ctx.Done():
			return
		case <-ticker.C:
		}
	}
}

func (r *Runtime) utcnow() time.Time {
	if r.now == nil {
		return time.Now().UTC()
	}
	return r.now().UTC()
}

func publicMessage(source string, err error) string {
	if err == nil {
		return source + " collection failed"
	}
	message := strings.TrimSpace(err.Error())
	if source == "zcode" {
		allowed := []string{
			"ZCode live Goal database read failed",
			"ZCode task index read failed",
			"Unsupported ZCode task index schema",
			"ZCode task sources are unavailable",
		}
		for _, value := range allowed {
			if message == value {
				return value
			}
		}
		return "ZCode collection failed"
	}

	allowedPrefixes := []string{
		"ZCode provider config could not be read",
		"CommandCode provider not found in ZCode",
		"CommandCode provider API key missing in ZCode",
		"CommandCode provider endpoint is not verified",
		"CommandCode billing request failed",
		"CommandCode billing response too large",
		"CommandCode authentication failed",
		"CommandCode billing access denied",
		"CommandCode billing API unavailable",
		"Unsupported CommandCode credits response",
		"Unsupported CommandCode subscription response",
	}
	for _, prefix := range allowedPrefixes {
		if strings.HasPrefix(message, prefix) {
			return truncateMessage(message)
		}
	}
	return "CommandCode collection failed"
}

func truncateMessage(message string) string {
	const limit = 240
	if len(message) <= limit {
		return message
	}
	return message[:limit]
}
